index.bs

```diff
@@ -931,6 +1019,231 @@ steps:
+1. Return |handle|.
+
+</div>
+
+### The {{FileSystemSyncAccessHandle/read()}} method ### {#api-filesystemsyncaccesshandle-read}
+
+<div class="note domintro">
+  : |handle| . {{FileSystemSyncAccessHandle/read()|read}}(|buffer|)
+  : |handle| . {{FileSystemSyncAccessHandle/read()|read}}(|buffer|, { {{FileSystemReadWriteOptions/at}} })
+  :: Reads the contents of the file associated with |handle| into |buffer|, optionally at a given offset.
+</div>
+
+<div algorithm>
+The <dfn method for=FileSystemSyncAccessHandle>read(|buffer|, {{FileSystemReadWriteOptions}}: |options|)</dfn> method, when invoked, must run
+these steps:
+
+1. If [=this=].[=[[state]]=] is "`closed`", throw an {{InvalidStateError}}.
+1. Let |bufferSize| be |buffer|'s [=byte length=].
+1. Let |fileContents| be [=this=].[=[[file]]=]'s [=file entry/binary data=].
+1. Let |fileSize| be |fileContents|'s [=byte sequence/length=].
+1. Let |readStart| be |options|.{{FileSystemReadWriteOptions/at}}.
+1. If |readStart| is larger than |fileSize|, return 0.
+1. Let |readEnd| be |readStart| + (|bufferSize| &minus; 1).
+1. If |readEnd| is larger than |fileSize|, set |readEnd| to |fileSize|.
+1. Let |result| be |bufferSize|.
+1. Let |bytes| be a [=byte sequence=] containing the bytes from |readStart| to |readEnd| of |fileContents|.
+1. If the operations reading from |fileContents| in the previous steps failed:
+    1. If there were partial reads  and the number of modified bytes in |bytes|  is known,
+       set |result| to the number of modified bytes.
+    1. Else set |result| to 0.
+1. Let |arrayBuffer| be |buffer|'s [=underlying buffer=].
+1. [=write|Write=] |bytes| into |arrayBuffer|.
+1. Return |result|.
+
+</div>
+
+### The {{FileSystemSyncAccessHandle/write()}} method ### {#api-filesystemsyncaccesshandle-write}
+
+<div class="note domintro">
+  : |handle| . {{FileSystemSyncAccessHandle/write()|write}}(|buffer|)
+  : |handle| . {{FileSystemSyncAccessHandle/write()|write}}(|buffer|, { {{FileSystemReadWriteOptions/at}} })
+  :: Writes the content of |buffer| into the file associated with |handle|, optionally at a given offset.
+</div>
+
+// TODO(fivedots): Figure out how to properly check the available storage quota (in this method and others) by passing the right storage shelf.
+// TODO(fivedots): Figure out mechanism to prevent concurrent IO operations e.g. doing write() while a truncate() is executing in parallel.
+<div algorithm>
+The <dfn method for=FileSystemSyncAccessHandle>write(|buffer|, {{FileSystemReadWriteOptions}}: |options|)</dfn> method, when invoked, must run
+these steps:
+
+1. If [=this=].[=[[state]]=] is "`closed`", throw a {{InvalidStateError}}.
+1. Let |writePosition| be |options|.{{FileSystemReadWriteOptions/at}}.
+1. Let |fileContents| be a copy of [=this=].[=[[file]]=]'s [=file entry/binary data=].
+1. Let |oldSize| be |fileContents|'s [=byte sequence/length=].
+1. Let |data| be [=get a copy of the buffer source|a copy of=] |buffer|.
```

Comment:
After thinking it through a bit more, I'm not sure the direct reading mechanism used in that algorithm will work for us.

The approach I was considering was: set the file's data to |head| + a sequence of 0 the length of buffer + |tail|, and then directly filling in the 0's with the contents of buffer. The problem is that, in reality, implementations are likely going to directly modify the affected section of the file. In the case of a partial write it would be hard/slow to comply with the spec (e.g. if there is a partial write during the direct reading of the buffer, a section of the file should be left with the original 0's).

I've changed the language to be more purposely vague, similar to what we'd have to with flush and other operations that are close to the OS. WDYT?

Reply:
", at which point write order and the management partial write cannot be generally specified." -- I think this needs to be reworded to be clearer (and "the management partial write" doesn't make syntactical sense)
" return the number of bytes that were written from data." (for partial writes): 'data' isn't defined here; I think you mean 'buffer'.
We may want to add a note to developers to always check the return from write() to detect failures, since we throw in most cases, but for partial writes we return a smaller size.   Perhaps we should throw here as well, and perhaps provide some way to know how much was written in the failure case.

Reply:
I clarified this section, added a note for web developers above, and fixed the |data|<->|buffer| issue.